### PR TITLE
Add azure functions to templates

### PR DESCRIPTION
**Pull request by jennyf19**
https://github.com/AzureAD/microsoft-identity-web/issues/899

Testing TODO from templates:
- [ ] With all the auth options 
  - [ ] dotnet new func2 --auth SingleOrg
  - [ ] dotnet new func2 --auth  IndividualB2C
  - [ ] dotnet new func2 --auth SingleOrg -calls-graph
  - [ ] dotnet new func2 --auth SingleOrg -calls-webapi
  - [ ] dotnet new func2
 
- [ ] Update wiki w/new template

### CODE REVIEW

**Comment by jennyf19** [comment]
@jmprieur next steps are to test the actual templates and update the .bat files.

File: ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            return new JsonResult(responseMessage);
+        }
+#else
+        public SampleFunc(ILogger<SampleFunc> logger)
+        {
+            _logger = logger;
+        }
+
+        [FunctionName("SampleFunc")]
+        public async Task<IActionResult> Run(
```
Comment (jmprieur) [inline ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs:108]: This method only needs to be async if `await req.HttpContext.AuthenticateAzureFunctionAsync();` is called (see below), that is `#if !NoAuth`

Do we want to have another big case

#elseif (!NoAuth)
// we would verify the token and the scopes, and return a Task<IActionResult>

#else
// no  verification of token and scope, return just a IActionResult

#endif

File: ProjectTemplates/templates/Functions-CSharp/.template.config/dotnetcli.host.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jmprieur) [inline ProjectTemplates/templates/Functions-CSharp/.template.config/dotnetcli.host.json:1]: You probably want to copy most of the content of https://github.com/AzureAD/microsoft-identity-web/blob/master/ProjectTemplates/templates/WebApi-CSharp/.template.config/dotnetcli.host.json, except: 
- UseLocalDB
- OrgReadAccess
- DisableOpenAPI

**Review by jmprieur — COMMENTED** [review]
This is a great start. Really impressive work, @jennyf19 !
I've left a few comments of things to improve.

I'll do another iteration when these are fixed.

File: ProjectTemplates/templates/Functions-CSharp/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+using Microsoft.Azure.Functions.Extensions.DependencyInjection;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+#if (GenerateGraph)
+using Microsoft.Graph;
+#endif
+
+[assembly: FunctionsStartup(typeof(SampleFunc.Startup))]
```
Comment (jmprieur) [inline ProjectTemplates/templates/Functions-CSharp/Startup.cs:15]:
```suggestion
[assembly: FunctionsStartup(typeof(Company.FunctionApp1.Startup))]
```

Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/Startup.cs:15]: nice catch!!!

File: ProjectTemplates/templates/Functions-CSharp/.template.config/template.json [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    "sources": [
+        {
+            "modifiers": [
+                {
+                    "condition": "(ExcludeLaunchSettings)",
+                    "exclude": [
+                        "Properties/launchSettings.json"
+                    ]
+                }
+            ]
```
[the remark this answers is not in the document]
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/.template.config/template.json:204]: yes, that's true i had started this section over. nice catch

---
In reply to: [570239757](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570239757) [](ancestors = 570239757)

File: ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req,
+            ILogger log)
+        {
+            _logger.LogInformation("C# HTTP trigger function processed a request.");
+
+            var (authenticationStatus, authenticationResponse) =
```
[the remark this answers is not in the document]
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs:116]: good catch. thanks. fixed.

---
In reply to: [570240657](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570240657) [](ancestors = 570240657)

File: ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs (hunk 1 above)
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs:110]: fixed.

---
In reply to: [570245685](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570245685) [](ancestors = 570245685)

File: ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+using System.Net.Http;
+#endif
+#if (GenerateGraph)
+using Microsoft.Graph;
+#endif
+using Microsoft.Extensions.Logging;
+#if (!NoAuth)
+using Microsoft.AspNetCore.Http;
+using Microsoft.Identity.Web.Resource;
+#endif
```
[the remark this answers is not in the document]
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs:16]: i see. thanks. fixed.

---
In reply to: [570242942](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570242942) [](ancestors = 570242942)

File: ProjectTemplates/templates/Functions-CSharp/.template.config/dotnetcli.host.json (hunk 2 above)
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/.template.config/dotnetcli.host.json:1]: yes, forgot this one. thanks. done.

---
In reply to: [570252768](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570252768) [](ancestors = 570252768)

File: tests/AzureFunctions/SampleFunc/SampleFunc.csproj [hunk 7]
```diff
@@ -1,22 +1,27 @@
...
     <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0" />
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\..\..\src\Microsoft.Identity.Web.MicrosoftGraph\Microsoft.Identity.Web.MicrosoftGraph.csproj" />
     <ProjectReference Include="..\..\..\src\Microsoft.Identity.Web\Microsoft.Identity.Web.csproj" />
   </ItemGroup>
   <ItemGroup>
-    <None Update="host.json">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    <None Update="appsettings.json">
```
Comment (jmprieur) [inline tests/AzureFunctions/SampleFunc/SampleFunc.csproj:20]: Content ?

File: ProjectTemplates/templates/Functions-CSharp/Company.FunctionApp1.csproj [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <None Update="host.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="local.settings.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
+    </None>
+  </ItemGroup>
```
Comment (jmprieur) [inline ProjectTemplates/templates/Functions-CSharp/Company.FunctionApp1.csproj:23]: It's missing the appsettings.json?

File: ProjectTemplates/templates/Functions-CSharp/.template.config/template.json [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+                {
+                    "condition": "(ExcludeLaunchSettings)",
+                    "exclude": [
+                        "Properties/launchSettings.json"
+                    ]
+                },
+                {
+                    "condition": "(NoAuth)",
+                    "exclude": [
+                        "Startup.cs"
```
Comment (jmprieur) [inline ProjectTemplates/templates/Functions-CSharp/.template.config/template.json:207]: We'll also want to exclude the `appsettings.json` (which is added by the `Startup.cs`)

File: tests/AzureFunctions/SampleFunc/SampleFunc.csproj (hunk 7 above)
Comment (jennyf19) [reply tests/AzureFunctions/SampleFunc/SampleFunc.csproj:20]: this is the dev app though...i'll check the template

---
In reply to: [570479260](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570479260) [](ancestors = 570479260)

File: ProjectTemplates/templates/Functions-CSharp/Company.FunctionApp1.csproj (hunk 8 above)
Comment (jennyf19) [reply ProjectTemplates/templates/Functions-CSharp/Company.FunctionApp1.csproj:23]: yes, fixed.

---
In reply to: [570481751](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570481751) [](ancestors = 570481751)

**Comment by jennyf19** [comment]
@jmprieur going to merge and then i'll do the items above.

**Pull request merged**